src/raft.c

```diff
@@ -1640,14 +1756,48 @@ void RaftReqFree(RaftReq *req)
         RedisModule_FreeThreadSafeContext(req->ctx);
         RedisModule_UnblockClient(req->client, NULL);
     }
+
     RedisModule_Free(req);
 }
 
-RaftReq *RaftReqInit(RedisModuleCtx *ctx, enum RaftReqType type)
+void blockedTimedOut(RedisModuleCtx *ctx, void *data)
 {
+    RedisRaftCtx *rr = &redis_raft;
+    RaftReq *req = data;
+
+    /* don't need to attach the req to this entry, as the req is part of the BlockedClient object */
+    raft_entry_t *entry = raft_entry_new(sizeof(raft_index_t));
+    entry->id = rand();
+    entry->type = RAFT_LOGTYPE_TIMEOUT_BLOCKED;
+    memcpy(entry->data, &req->raft_idx, entry->data_len);
+
+    int e = raft_recv_entry(rr->raft, entry, NULL);
+    if (e != 0) {
+        /* if we timed out, but we couldn't apply the timeout entry, it means we lost leadership.
+         * Therefore, when we apply a new NO_OP, it will automatically be timed out on all nodes.
+         * Therefore, don't have to do anything here
+         */
+    }
+
+    raft_entry_release(entry);
+}
+
+/* timeout is used to determine if there is a timeout and if it's a blocking command.
+ * non blocking commands should use -1.  Blocking commands should use 0 if there's no timeout, or
+ * a positive number if there is a timeout
+ */
+RaftReq *RaftReqInit(RedisModuleCtx *ctx, enum RaftReqType type, long long timeout)
+{
+    bool blocking = timeout >= 0 ? true : false;
+
     RaftReq *req = RedisModule_Calloc(1, sizeof(RaftReq));
     if (ctx != NULL) {
-        req->client = RedisModule_BlockClient(ctx, NULL, NULL, NULL, 0);
+        req->client_id = RedisModule_GetClientId(ctx);
+        if (blocking) {
+            req->client = RedisModule_BlockClient(ctx, NULL, NULL, NULL, 0);
```

Comment:
There is no different in the behavior between the two cases...

Reply:
as of now there isn't, it depends if we need a timeout callback for client unblock.

Reply:
I removed it for now with my just pushed change for RaftReqInit()